src/xref2/link.ml

```diff
@@ -28,6 +28,40 @@ let synopsis_of_module env (m : Component.Module.t) =
       | Ok sg -> synopsis_from_comment (Component.extract_signature_doc sg)
       | Error _ -> None)
 
+(** Duplicate labels are disambiguated by adding a "_2" suffix to them. The
+    order in which it is inserted into the env is used to decide which prefix to
+    assign, the first is unchanged. *)
+let ensure_label_unique env heading =
+  let rec disambiguate_label index name =
+    let new_name = name ^ "_" ^ string_of_int (index + 1) in
+    (* If ambiguous again, insert an extra '_' before the index. *)
+    match Env.lookup_by_name Env.s_label new_name env with
+    | Ok _ | Error (`Ambiguous _) -> disambiguate_label index (name ^ "_")
+    | Error `Not_found -> new_name
+  in
+  let location_equal a b = Location_.location a = Location_.location b in
+  (* Returns [0] if [heading] is the first occurence. The first occurence is at
+     the end of the list stored in [env]. Compare only the location as it is
+     unique to each heading and unlikely to be updated. *)
+  let rec dup_index = function
```

Comment:
Given that a lot of this code is removed in a subsequent commit, can we squash them?

Reply:
I've squashed the related commits.